packages/checkout/widgets-lib/src/widgets/sale/hooks/useCurrency.ts

```diff
@@ -0,0 +1,170 @@
+/* eslint-disable @typescript-eslint/naming-convention */
+import { useState, useEffect } from 'react';
+import { Passport } from '@imtbl/passport';
+
+import { Environment } from '@imtbl/config';
+import { TokenFilterTypes } from '@imtbl/checkout-sdk';
+import { providers } from 'ethers';
+import { PRIMARY_SALES_API_BASE_URL } from '../utils/config';
+
+import {
+  ClientConfig,
+  ClientConfigCurrency,
+  ClientConfigCurrencyConversion,
+  SaleWidgetCurrency,
+  SaleWidgetCurrencyType,
+  SignPaymentTypes,
+} from '../types';
+import { useCheckoutClient } from './useCheckoutClient';
+import { sortCurrencies } from '../utils/sortCurrencies';
+
+type ClientConfigResponse = {
+  contract_id: string;
+  currencies: {
+    base: boolean;
+    decimals: number;
+    erc20_address: string;
+    exchange_id: string;
+    name: string;
+  }[];
+  currency_conversion: {
+    [key: string]: {
+      amount: number;
+      name: string;
+      type: string;
+    };
+  };
+};
+
+const toClientConfig = (response: ClientConfigResponse): ClientConfig => ({
```

Comment:
since this logic became complex, I'll suggest we move into using pattern of creating utility functions within the `/functions` directory. And start adding some test to each utility.

example
https://github.com/immutable/ts-immutable-sdk/blob/853aaee1eb1b00f61d95735576e8b1c8006f949b/packages/checkout/widgets-lib/src/widgets/swap/functions/FetchQuote.ts